File: apps/demo-be-elixir-phoenix/lib/demo_be_exph/expense/expense.ex
defmodule DemoBeExph.Expense.Expense do
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:id, :binary_id, autogenerate: true}
  @foreign_key_type :binary_id
  @timestamps_opts [inserted_at: :created_at, updated_at: :updated_at, type: :utc_datetime]

  @supported_currencies ~w(USD IDR)
  @supported_units ~w(liter ml kg g km meter gallon lb oz mile piece hour)
  @supported_types ~w(income expense)

  schema "expenses" do
    field :user_id, :binary_id
    field :amount, :decimal
    field :currency, :string
    field :category, :string
    field :type, :string
    field :description, :string
    field :unit, :string
    field :quantity, :decimal
    field :date, :date
    field :created_by, :string
    field :updated_by, :string
    field :deleted_at, :utc_datetime
    field :deleted_by, :string

    timestamps()
  end

  def changeset(expense, attrs) do
    expense
    |> cast(attrs, [
      :user_id,
      :amount,
      :currency,
      :category,
      :type,
      :description,
      :unit,
      :quantity,
      :date
    ])
    |> update_change(:type, &String.downcase/1)
    |> validate_required([:user_id, :amount, :currency, :category, :type, :description, :date])
    |> validate_inclusion(:currency, @supported_currencies,
      message: "is not supported. Supported: #{Enum.join(@supported_currencies, ", ")}"
    )
    |> validate_inclusion(:type, @supported_types, message: "must be 'income' or 'expense'")
    |> validate_number(:amount,
      greater_than: Decimal.new(0),
      message: "must be greater than 0"
    )
    |> validate_unit()
  end

  def update_changeset(expense, attrs) do
    expense
    |> cast(attrs, [:amount, :currency, :category, :type, :description, :unit, :quantity, :date])
    |> update_change(:type, &String.downcase/1)
    |> validate_required([:amount, :currency, :category, :type, :description, :date])
    |> validate_inclusion(:currency, @supported_currencies,
      message: "is not supported. Supported: #{Enum.join(@supported_currencies, ", ")}"
    )
    |> validate_inclusion(:type, @supported_types, message: "must be 'income' or 'expense'")
    |> validate_number(:amount,
      greater_than: Decimal.new(0),
      message: "must be greater than 0"
    )
    |> validate_unit()
  end

  defp validate_unit(changeset) do
    unit = get_change(changeset, :unit)

    if unit && unit not in @supported_units do
      add_error(
        changeset,
        :unit,
        "is not supported. Supported: #{Enum.join(@supported_units, ", ")}"
      )
    else
      changeset
    end
  end
end

File: apps/crud-be-fsharp-giraffe/src/DemoBeFsgi/Domain/Expense.fs
module DemoBeFsgi.Domain.Expense

open System
open DemoBeFsgi.Domain.Types

type Expense =
    { Id: Guid
      UserId: Guid
      Amount: decimal
      Currency: string
      Category: string
      Description: string
      Date: DateTime
      Type: string // "income" or "expense"
      Quantity: decimal option
      Unit: string option
      CreatedAt: DateTime
      UpdatedAt: DateTime }

let validateAmount (amount: decimal) =
    if amount < 0m then
        Error(ValidationError("amount", "Amount must not be negative"))
    else
        Ok amount

let validateCurrencyPrecision (currency: string) (amount: decimal) =
    match currency.ToUpperInvariant() with
    | "USD" ->
        if amount <> Math.Round(amount, 2) then
            Error(ValidationError("amount", "USD requires at most 2 decimal places"))
        else
            Ok amount
    | "IDR" ->
        if amount <> Math.Round(amount, 0) then
            Error(ValidationError("amount", "IDR requires 0 decimal places"))
        else
            Ok amount
    | _ -> Error(ValidationError("currency", $"Unsupported currency: %s{currency}"))

let validateUnit (unit: string option) =
    match unit with
    | None
    | Some "" -> Ok unit
    | Some u ->
        if supportedUnits.Contains(u.ToLowerInvariant()) then
            Ok(Some(u.ToLowerInvariant()))
        else
            Error(ValidationError("unit", $"Unsupported unit: %s{u}"))
